javascripts/2048.js

```diff
@@ -1,32 +1,347 @@
+        col2.push(tile[i]);
+      } else if (tile[i].getAttribute("data-col") == "c3") {
+        col3.push(tile[i]);
+      }
+    }
+
+  var cols = [col0, col1, col2, col3];
+
+  return cols;
+}
+
+function moveColUp(col) {
+  for (var i = 0; i < col.length; i++) {
+    for(var j = 0; j < col[i].length; j++) {
+      col[i][j].setAttribute("data-row", ("r" + j));
+    }
+  }
+}
+
+function collideTilesUp(cols){
+
+  for (var i = 0; i < cols.length; i++) {
+    for (var j = 0; j< cols[i].length-1; j++){
+      var upTileValue = cols[i][j].getAttribute("data-val");
+      var nextTileValue = cols[i][j+1].getAttribute("data-val");
+      if (upTileValue == nextTileValue){
+        var newTileValue = upTileValue * 2;
+        cols[i][j].setAttribute("data-val", newTileValue);
+        cols[i][j].textContent = newTileValue;
+        cols[i][j+1].remove();
+    };
+
+  }
+}
+}
+
+function moveColDown(col) {
+  for (var i = 0; i < col.length; i++) {
+    for(var j = 0; j < col[i].length; j++) {
+      col[i][j].setAttribute("data-row",("r" + (j + 4 - col[i].length)));
+    }
+  }
+}
+
+function collideTilesDown(cols){
+  for (var i = 0; i < cols.length; i++) {
+
+  for (var j = 0; j < cols[i].length -1; j++){
+   var downTileValue = cols[i][j].getAttribute("data-val");
+   var nextTileValue = cols[i][j+1].getAttribute("data-val");
+      if (downTileValue == nextTileValue){
+        var newTileValue = downTileValue * 2;
+          cols[i][j].setAttribute("data-val", newTileValue);
+          cols[i][j].textContent = newTileValue;
+          cols[i][j+1].remove();
+    };
+  }
+}
+}
+
+
+function gameLost(){
+  var gameLost = false;
+
+  if (($("#gameboard .tile").length == 16) ) {  // && (no collisions possible)
+    var gameLost = true;
+  }
+
+ return gameLost;
+}
+
+function gameWon(){
+  var gameWon = false;
+
+  var gameBoard = $("#gameboard .tile")
+  for (var i = 0; i < gameBoard; i++){  // loop through all the tiles, check value
+     if (gameBoard[i].attr("data-val") >= 2048) {    // if tile with value "2048", gameWon true
+        gameWon = true;
+      }
+    }
+  if (gameWon == true) {
+    // $('body').keydown(function(event){
+    //   var arrow_keys = [37, 38, 39, 40];
+    //   if(arrow_keys.indexOf(event.which) > -1) {
+    //     $("body").fadeOut("slow");
+    //     event.preventDefault();
+    //   }
+    // });
+  }
+}
+
+function freeSpot() {
+ var freeSpots = [];
+ for (var row = 0; row < 4; row++) {
+   for (var col = 0; col < 4; col++) {
+     var reference = ".tile[data-row=r" + row + "][data-col=c" + col + "]";
+     // => .tile[data-row=r0][data-col=c0] => occupied
+     var unoccupied = $(reference).length == 0;
+     if(unoccupied)
+       freeSpots.push({ row: row, col: col });
+   }
+ }
+
+ return freeSpots[Math.floor( Math.random() * freeSpots.length )]; // => returns hash
+}
+
+function addTile(){
+  var newSpot = freeSpot();
+  var newDiv = document.createElement("div");
```

Comment:
Interesting.  Any particular reason you went with native DOM functions instead of using jQuery here?